numba/targets/linalg.py

```diff
@@ -1574,3 +1574,112 @@ def pinv_impl(a, rcond=1.e-15):
         return acpy.T.ravel().reshape(a.shape).T
 
     return pinv_impl
+
+# Walks the diag of a LUP decomposed matrix
+# uses that det(A) = prod(diag(lup(A)))
+# and also that log(a)+log(b) = log(a*b)
+# The return sign is adjusted based on the values found
+# such that the log(value) stays in the real domain.
+
+
+def _get_slogdet_diag_walker(a):
+    if isinstance(a.dtype, types.scalars.Complex):
+        @jit(nopython=True)
+        def cmplx_diag_walker(n, a, sgn):
+            # walk diagonal
+            csgn = sgn + 0.j
+            acc = 0.
+            for k in range(n):
+                absel = np.abs(a[k, k])
+                csgn = csgn * (a[k, k] / absel)
+                acc = acc + np.log(absel)
+            return (csgn, acc)
+        return cmplx_diag_walker
+    else:
+        @jit(nopython=True)
+        def real_diag_walker(n, a, sgn):
+            # walk diagonal
+            acc = 0.
+            for k in range(n):
+                v = a[k, k]
+                if v < 0.:
+                    sgn = -sgn
+                    v = -v
+                acc = acc + np.log(v)
+            # sgn is a float dtype
+            return (sgn + 0., acc)
+        return real_diag_walker
+
+
+@overload(np.linalg.slogdet)
+def slogdet_impl(a):
+    ensure_lapack()
+
+    _check_linalg_matrix(a, "slogdet")
+
+    numba_xxgetrf_sig = types.intc(types.char,   # kind
+                                   types.intp,  # m
+                                   types.intp,  # n
+                                   types.CPointer(a.dtype),  # a
+                                   types.intp,  # lda
+                                   types.CPointer(F_INT_nbtype)  # ipiv
+                                   )
+
+    numba_xxgetrf = types.ExternalFunction("numba_xxgetrf",
+                                           numba_xxgetrf_sig)
+
+    kind = ord(get_blas_kind(a.dtype, "slogdet"))
+
+    F_layout = a.layout == 'F'
+
+    diag_walker = _get_slogdet_diag_walker(a)
+
+    def slogdet_impl(a):
+        n = a.shape[-1]
+        if a.shape[-2] != n:
+            msg = "Last 2 dimensions of the array must be square."
+            raise np.linalg.LinAlgError(msg)
+
+        _check_finite_matrix(a)
+
+        if F_layout:
+            acpy = np.copy(a)
+        else:
+            acpy = np.asfortranarray(a)
+
+        ipiv = np.empty(n, dtype=F_INT_nptype)
+
+        r = numba_xxgetrf(kind, n, n, acpy.ctypes, n, ipiv.ctypes)
+
+        if r > 0:
+            # factorisation failed, return same defaults as np
+            return (0., -np.inf)
+        _inv_err_handler(r)  # catch input-to-lapack problem
+
+        sgn = 1
+        for k in range(n):
+            sgn = sgn + (ipiv[k] != (k + 1))
```

Comment:
Nothing I can find directly. Suppose there were no row interchanges taking place and the target was to return the determinant itself, the determinant could then just be calculated directly as the product of the diagonal values of the decomposition. However, as this needs to return the sign and log of the determinant we use that `log(a*b)=log(a)+log(b)` so the determinant is the sum of the `log` of the values on the diagonal but obviously the sign needs to be accounted for. Recalling multiplication order matters to the sign, assuming again there's no pivoting, the `_get_slogdet_diag_walker` executes this idea (make values positive if they aren't already and adjust the accumulated sign, then accumulate the log value). Finally, if we introduce partial pivoting with row interchanges as in the LAPACK routine above, the above line and the modulo operator is deciding if the number of interchanges performed is odd or even, this taking into account the affect of the permutation on the multiplication order over the diagonal, this sign is then carried forward into the `_get_slogdet_diag_walker` routine and we can pretend the permutations didn't happen knowing the adjustment has been made.